### PR TITLE
`generator` WIP

### PR DESCRIPTION
**Pull request by CaseyCarter**
My reference implementation of `generator` from the proposal. This needs more thorough testing, a quick audit to ensure it meets all the requirements, and a review of the allocator handling. I haven't had time to work on this, so I'm putting it in a feature branch where others can contribute if they like.

Works towards #2936.

### CODE REVIEW

**Comment by CaseyCarter** [comment]
Bah, I need to audit this for unnecessary clang-format suppression.

File: stl/inc/ranges [hunk 1]
```diff
@@ -77,6 +77,21 @@ namespace ranges {
     template <class _Ty, size_t _Extent>
         requires (_Extent != dynamic_extent)
     inline constexpr auto _Compile_time_max_size<const span<_Ty, _Extent>> = _Extent;
+
+#ifdef __cpp_lib_byte
+    using _Elements_alloc_type = byte;
+#else
+    using _Elements_alloc_type = char;
+#endif
```
Comment (frederick-vs-ja) [inline stl/inc/ranges:85]: Hmm, perhaps we should be consistent on whether there's workaround when `__cpp_lib_byte` is not defined, see https://github.com/microsoft/STL/pull/4337#discussion_r1462569445.

**Comment by StephanTLavavej** [comment]
Clang x86 failures:

```
fatal error: error in backend: failed to perform tail call elimination on a call site marked musttail
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
```

File: stl/inc/ranges (hunk 1 above)
Comment (StephanTLavavej) [reply stl/inc/ranges:85]: Yeah, this is different than our previous `__cpp_lib_byte` checks, and is ODR-violation-ish.

Really we need to push for the WinSDK to be fixed, but in the meantime we should definitely think about what to do here.

**Comment by CaseyCarter** [comment]
> Clang x86 failures

If you're adding workarounds for one compiler bug while removing workarounds for a different compiler bug, you may be a standard library implementor.

**Comment by cpplearner** [comment]
> Clang x86 failures:
> 
> ```
> fatal error: error in backend: failed to perform tail call elimination on a call site marked musttail
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
> ```

Seems like LLVM-56507?

**Comment by StephanTLavavej** [comment]
Ha, I filed that a year and a half ago and forgot! The bug report is coming from inside the house! :joy_cat:

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! Checks are green, and I did an extremely cursory review to confirm that all newly added files have our copyright/license banner, so let's get this party started. I also updated the tracking issue to capture your todos.

**Pull request merged**